# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1699464351 -3600
# Parent  2195a7850c64f91ce3280dc1e74111b986ea8d0b
Misc: Add update check and fix release notes link.
* * *
Misc: Startup update check.
* * *
Misc: Add more logging to update check.
* * *
Misc: CSS tweak so download link aligns correctly.
* * *
Misc: Improve behaviour when update check fails.
* * *
Misc: Make update check take policy DisableAppUpdate into account.

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -34,16 +34,17 @@ pref("app.update.checkInstallTime", true
 pref("app.update.timerMinimumDelay", 120);
 
 // The minimum delay in milliseconds for the first firing after startup of the timer
 // to notify consumers of the timer manager.
 // minimum=10 seconds, default=30 seconds, and maximum=120 seconds
 pref("app.update.timerFirstInterval", 30000);
 
 // App-specific update preferences
+pref("betterbird.update.checkIntervalDays", 5);
 
 // The interval to check for updates (app.update.interval) is defined in
 // the branding files.
 
 // Enables some extra Application Update Logging (can reduce performance)
 pref("app.update.log", false);
 // Causes Application Update Logging to be sent to a file in the profile
 // directory. This preference is automatically disabled on application start to
@@ -110,17 +111,17 @@ pref("app.update.service.enabled", true)
 #endif
 
 #ifdef XP_WIN
 // This pref prevents BITS from being used by Thunderbird to download updates.
 pref("app.update.BITS.enabled", false);
 #endif
 
 // Release notes URL
-pref("app.releaseNotesURL", "https://live.thunderbird.net/%APP%/releasenotes?locale=%LOCALE%&version=%VERSION%&channel=%CHANNEL%&os=%OS%&buildid=%APPBUILDID%");
+pref("app.releaseNotesURL", "https://www.betterbird.eu/releasenotes?locale=%LOCALE%&version=%VERSION%&channel=%CHANNEL%&os=%OS%&buildid=%APPBUILDID%");
 
 #ifdef XP_MACOSX
   // If set to true, Thunderbird will automatically restart if it is left
   // running with no windows open.
   pref("app.update.noWindowAutoRestart.enabled", true);
   // How long to wait after all windows are closed before restarting,
   // in milliseconds. 5 min = 300000 ms.
   pref("app.update.noWindowAutoRestart.delayMs", 300000);
diff --git a/mail/base/content/aboutDialog.js b/mail/base/content/aboutDialog.js
--- a/mail/base/content/aboutDialog.js
+++ b/mail/base/content/aboutDialog.js
@@ -17,17 +17,17 @@ if (AppConstants.MOZ_UPDATER) {
 }
 
 window.addEventListener("DOMContentLoaded", onLoad);
 if (AppConstants.MOZ_UPDATER) {
   // This method is in the aboutDialog-appUpdater.js file.
   window.addEventListener("unload", onUnload);
 }
 
-function onLoad(event) {
+async function onLoad(event) {
   if (event.target !== document) {
     return;
   }
 
   let defaults = Services.prefs.getDefaultBranch(null);
   let distroId = defaults.getCharPref("distribution.id", "");
   if (distroId) {
     let distroAbout = defaults.getStringPref("distribution.about", "");
@@ -115,16 +115,87 @@ function onLoad(event) {
   }
   // Open internal (about:) links open in Thunderbird tab
   for (const link of document.getElementsByClassName("tab-link")) {
     link.onclick = event => {
       event.preventDefault();
       openAboutTab(event.target.href);
     };
   }
+
+  await BetterbirdUpdateCheck(window);
+}
+
+async function BetterbirdUpdateCheck(win) {
+  let needsUpdate = "arguments" in win && win.arguments[0].needsUpdate;
+  let checkForUpdate = !needsUpdate;
+  let cantCheck = false;
+  if (Services.policies && !Services.policies.isAllowed("appUpdate")) {
+    console.info(
+      "Betterbird skipping update check due to policy DisableAppUpdate"
+    );
+    checkForUpdate = false;
+    cantCheck = true;
+  }
+
+  if (checkForUpdate) {
+    try {
+      let response = await fetch(
+        "https://www.betterbird.eu/start/versions.txt",
+        { cache: "no-store" }
+      );
+      if (!response.ok) {
+        console.error("Betterbird update check FAILED");
+        cantCheck = true;
+      } else {
+        let content = await response.text();
+        let lines = content.split("\n");
+        let version = Services.appinfo.version;
+        let buildID = Services.appinfo.appBuildID;
+        let os = Services.appinfo.OS;
+        for (let l of lines) {
+          let parts = l.split(",");
+          // Format is os,version,buildid.
+          if (version.startsWith(parts[1]) && os == parts[0]) {
+            if (parts[2] > buildID) {
+              needsUpdate = true;
+              console.info(
+                `Betterbird update available for ${os}, new build ${parts[2]}, current build ${buildID}`
+              );
+            } else {
+              console.info(
+                `Betterbird update-to-date for ${os}, published build ${parts[2]}, current build ${buildID}`
+              );
+            }
+            break;
+          }
+        }
+      }
+    } catch (ex) {
+      console.error("Betterbird update check FAILED");
+      cantCheck = true;
+    }
+  }
+
+  if (needsUpdate) {
+    let el = document.getElementById("manualUpdate");
+    el.style = "display: normal";
+    // Replacement for
+    // https://searchfox.org/comm-central/rev/16c0bb426b179283b0b8a3076b4bf03208207c4a/mozilla/browser/base/content/aboutDialog-appUpdater.js#55
+    let link = document.getElementById("manualLink");
+    link.href = "https://www.betterbird.eu/downloads/";
+    link.textContent = "www.betterbird.eu/downloads";
+  } else if (cantCheck) {
+    let el = document.getElementById("checkingFailed");
+    el.style = "display: normal";
+  } else {
+    // Up to date.
+    let el = document.getElementById("noUpdatesFound");
+    el.style = "display: normal";
+  }
 }
 
 // This function is used to open about: tabs. The caller should ensure the url
 // is only an about: url.
 function openAboutTab(url) {
   // Check existing windows
   let mailWindow = Services.wm.getMostRecentWindow("mail:3pane");
   if (mailWindow) {
diff --git a/mail/base/content/aboutDialog.xhtml b/mail/base/content/aboutDialog.xhtml
--- a/mail/base/content/aboutDialog.xhtml
+++ b/mail/base/content/aboutDialog.xhtml
@@ -39,16 +39,28 @@
              data-l10n-id="release-notes-link"></a>
         </xul:hbox>
 
         <span id="distribution" class="text-blurb"></span>
         <span id="distributionId" class="text-blurb"></span>
 
         <xul:vbox id="detailsBox">
           <xul:hbox id="updateBox">
+            <div id="noUpdatesFound" style="display: none;">
+              <span data-l10n-id="update-no-updates-found"></span>
+            </div>
+            <div id="checkingFailed" style="display: none;">
+              <span data-l10n-id="aboutdialog-update-checking-failed"></span>
+            </div>
+            <div id="manualUpdate" style="display: none;">
+              <span data-l10n-id="update-manual">
+                <a id="manualLink" data-l10n-name="manual-link"
+                   class="manualLink text-link browser-link"></a>
+              </span>
+            </div>
 #ifdef MOZ_UPDATER
             <div id="icons">
               <img class="icon update-throbber" role="presentation"/>
               <img class="icon noUpdatesFound" src="chrome://global/skin/icons/check.svg" role="presentation"/>
               <img class="icon apply" src="chrome://global/skin/icons/reload.svg" role="presentation"/>
             </div>
             <xul:vbox>
               <xul:deck id="updateDeck" orient="vertical">
diff --git a/mail/components/MailGlue.jsm b/mail/components/MailGlue.jsm
--- a/mail/components/MailGlue.jsm
+++ b/mail/components/MailGlue.jsm
@@ -694,16 +694,17 @@ MailGlue.prototype = {
    * _scheduleBestEffortUserIdleTasks.
    * Don't be fooled by thinking that the use of the timeout parameter
    * will delay your function: it will just ensure that it potentially
    * happens _earlier_ than expected (when the timeout limit has been reached),
    * but it will not make it happen later (and out of order) compared
    * to the other ones scheduled together.
    */
   _scheduleStartupIdleTasks() {
+    let self = this;
     const idleTasks = [
       {
         task() {
           // This module needs to be loaded so it registers to receive
           // FormAutoComplete:GetSelectedIndex messages and respond
           // appropriately, otherwise we get error messages like the one
           // reported in bug 1635422.
           ChromeUtils.importESModule(
@@ -772,16 +773,21 @@ MailGlue.prototype = {
       {
         condition: AppConstants.MOZ_UPDATER,
         task: () => {
           lazy.UpdateListener.maybeShowUnsupportedNotification();
         },
       },
       {
         task() {
+          self.BetterbirdUpdateCheck();
+        },
+      },
+      {
+        task() {
           // Use idleDispatch a second time to run this after the per-window
           // idle tasks.
           ChromeUtils.idleDispatch(() => {
             Services.obs.notifyObservers(
               null,
               "mail-startup-idle-tasks-finished"
             );
           });
@@ -887,16 +893,116 @@ MailGlue.prototype = {
           type: "contentTab",
           tabParams,
         }
       );
       linkHandled.data = true;
     }
   },
 
+  async BetterbirdUpdateCheck() {
+    if (Services.policies && !Services.policies.isAllowed("appUpdate")) {
+      console.info(
+        "Betterbird skipping update check due to policy DisableAppUpdate"
+      );
+      return;
+    }
+
+    let lastCheckTime = Services.prefs.getIntPref(
+      "betterbird.update.lastCheckTime",
+      0
+    );
+    let now = Math.round(Date.now() / 1000);
+    let days = Services.prefs.getIntPref(
+      "betterbird.update.checkIntervalDays",
+      5
+    );
+    if (days < 0) {
+      days = 5;
+    } else if (days == 0) {
+      console.info("Betterbird skipping update check, interval set to zero");
+      return;
+    }
+    let checkInterval = days * 24 * 60 * 60;
+    if (now - lastCheckTime < checkInterval) {
+      console.info(
+        `Betterbird skipping update check, last done ${new Date(
+          lastCheckTime * 1000
+        ).toLocaleString()}, interval ${days} days`
+      );
+      return;
+    }
+    Services.prefs.setIntPref("betterbird.update.lastCheckTime", now);
+    let needsUpdate = false;
+    try {
+      let response = await fetch(
+        "https://www.betterbird.eu/start/versions.txt",
+        { cache: "no-store" }
+      );
+      if (!response.ok) {
+        console.error(
+          `Betterbird update check FAILED, last done ${new Date(
+            lastCheckTime * 1000
+          ).toLocaleString()}`
+        );
+        return;
+      }
+      let content = await response.text();
+      let lines = content.split("\n");
+      let version = Services.appinfo.version;
+      let buildID = Services.appinfo.appBuildID;
+      let os = Services.appinfo.OS;
+      for (let l of lines) {
+        let parts = l.split(",");
+        // Format is os,version,buildid.
+        if (version.startsWith(parts[1]) && os == parts[0]) {
+          if (parts[2] > buildID) {
+            needsUpdate = true;
+            console.info(
+              `Betterbird update available for ${os}, new build ${parts[2]}, current build ${buildID}`
+            );
+          } else {
+            console.info(
+              `Betterbird update-to-date for ${os}, published build ${parts[2]}, current build ${buildID}`
+            );
+          }
+          break;
+        }
+      }
+    } catch (ex) {
+      console.error(
+        `Betterbird update check FAILED, last done ${new Date(
+          lastCheckTime * 1000
+        ).toLocaleString()}`
+      );
+      return;
+    }
+    if (!needsUpdate) {
+      return;
+    }
+    let win = Services.wm.getMostRecentWindow("mail:3pane");
+    if (!win) {
+      return;
+    }
+    let features = "chrome,";
+    if (AppConstants.platform == "win") {
+      features += "centerscreen,dependent";
+    } else if (AppConstants.platform == "macosx") {
+      features += "resizable=no,minimizable=no";
+    } else {
+      features += "centerscreen,dependent,dialog=no";
+    }
+    win.openDialog(
+      "chrome://messenger/content/aboutDialog.xhtml",
+      "",
+      features,
+      { needsUpdate }
+    );
+  },
+
   // for XPCOM
   QueryInterface: ChromeUtils.generateQI(["nsIObserver"]),
 };
 
 /**
  * Report account types to telemetry. For im accounts, use `im_protocol` as
  * scalar key name.
  */
